snuba/snuba_migrations/replays/0019_add_materialization.py

```diff
@@ -0,0 +1,192 @@
+    anyIfState(environment, environment != '') as environment,
+    anyState(ip_address_v4) as ip_address_v4,
+    anyState(ip_address_v6) as ip_address_v6,
+    sumState(toUInt64(is_archived)) as is_archived,
+    anyIfState(os_name, os_name != '') as os_name,
+    anyIfState(os_version, os_version != '') as os_version,
+    anyIfState(platform, platform != '') as platform,
+    any(retention_days),
+    anyIfState(sdk_name, sdk_name != '') as sdk_name,
+    anyIfState(sdk_version, sdk_version != '') as sdk_version,
+    minState(replay_start_timestamp) as started_at,
+    anyIfState(user, user != '') as user,
+    anyIfState(user_id, user_id != '') as user_id,
+    anyIfState(user_name, user_name != '') as user_name,
+    anyIfState(user_email, user_email != '') as user_email,
+    minState(segment_id) as min_segment_id
+FROM replays_local
+GROUP BY project_id, toStartOfHour(timestamp), replay_id
+""",
+        target=operations.OperationTarget.LOCAL,
+    )
+
+
+def backward_iter() -> Iterator[operations.SqlOperation]:
+    yield operations.DropTable(
+        storage_set=StorageSetKey.REPLAYS,
+        table_name="replays_aggregation_mv",
+        target=operations.OperationTarget.LOCAL,
+    )
+
+    yield operations.DropTable(
+        storage_set=StorageSetKey.REPLAYS,
+        table_name="replays_aggregated_local",
+        target=operations.OperationTarget.LOCAL,
+    )
+
+
+def any_if_string(
+    column_name: str, nullable: bool = False, low_cardinality: bool = False
+) -> Column[Modifiers]:
+    return Column(
+        column_name,
+        AggregateFunction(
+            "anyIf",
+            [
+                String(Modifiers(nullable=nullable, low_cardinality=low_cardinality)),
+                UInt(8, Modifiers(nullable=nullable)),
+            ],
+        ),
+    )
+
+
+def any_if_nullable_string(
+    column_name: str, low_cardinality: bool = False
+) -> Column[Modifiers]:
+    """Returns an aggregate anyIf function."""
+    return any_if_string(column_name, nullable=True, low_cardinality=low_cardinality)
+
+
+def sum(column_name: str) -> Column[Modifiers]:
+    """Returns an aggregate sum function."""
+    return Column(column_name, AggregateFunction("sum", [UInt(64)]))
+
+
+def count_nullable(column_name: str) -> Column[Modifiers]:
+    """Returns an aggregate count function capable of accepting nullable integer values."""
+    return Column(
+        column_name, AggregateFunction("count", [UInt(64, Modifiers(nullable=True))])
+    )
+
+
+columns: List[Column[Modifiers]] = [
+    # Primary-key.
+    Column("project_id", UInt(64)),
+    Column("to_hour_timestamp", DateTime()),
+    Column("replay_id", UUID()),
+    # Columns ordered by column-name.
+    any_if_nullable_string("browser_name"),
+    any_if_nullable_string("browser_version"),
+    sum("count_dead_clicks"),
+    sum("count_errors"),
+    sum("count_infos"),
+    sum("count_rage_clicks"),
+    count_nullable("count_segments"),
+    sum("count_urls"),
```

Comment:
@cmanallen I see that some of these columns are materialized already, so with this materialized table, are we now going to be doing materialization twice? would we need the materialized columns still if we have this table? maybe I don't quite know exactly from the product standpoint if both are needed.

Reply:
@MeredithAnya We'll still need the materialized columns.  Some queries aren't possible with the materialized view so we have a tiered system where performance degrades as queries get more complex.